ntcore/src/main/native/cpp/local/LocalStorageImpl.cpp

```diff
@@ -0,0 +1,1086 @@
+        network->ClientPublish(Handle{publisher->handle}.GetIndex(),
+                               topic->name, topic->typeStr, topic->properties,
+                               publisher->config);
+        anyPublisher = publisher;
+      }
+    }
+    if (anyPublisher && topic->lastValue) {
+      network->ClientSetValue(Handle{anyPublisher->handle}.GetIndex(),
+                              topic->lastValue);
+    }
+  }
+  for (auto&& subscriber : m_subscribers) {
+    if (!subscriber->config.hidden) {
+      network->ClientSubscribe(1 + Handle{subscriber->handle}.GetIndex(),
+                               {{subscriber->topic->name}}, subscriber->config);
+    }
+  }
+  for (auto&& subscriber : m_multiSubscribers) {
+    if (!subscriber->options.hidden) {
+      network->ClientSubscribe(-1 - Handle{subscriber->handle}.GetIndex(),
+                               subscriber->prefixes, subscriber->options);
+    }
+  }
+}
+
+void StorageImpl::ClearNetwork() {
+  DEBUG4("ClearNetwork()");
+  m_network = nullptr;
+  // treat as an unannounce all from the network side
+  for (auto&& topic : m_topics) {
+    RemoveNetworkPublisher(topic.get());
+  }
+}
+
+//
+// Topic functions
+//
+
+LocalTopic* StorageImpl::GetOrCreateTopic(std::string_view name) {
+  auto& topic = m_nameTopics[name];
+  // create if it does not already exist
+  if (!topic) {
+    topic = m_topics.Add(m_inst, name);
+    // attach multi-subscribers
+    for (auto&& sub : m_multiSubscribers) {
+      if (sub->Matches(name, topic->special)) {
+        topic->multiSubscribers.Add(sub.get());
+      }
+    }
+  }
+  return topic;
+}
+
+//
+// Topic property functions
+//
+
+void StorageImpl::SetFlags(LocalTopic* topic, unsigned int flags) {
+  wpi::json update = topic->SetFlags(flags);
+  if ((flags & NT_UNCACHED) != 0 && (flags & NT_PERSISTENT) != 0) {
+    WARN("topic {}: disabling cached property disables persistent storage",
+         topic->name);
+  }
+  if (!update.empty()) {
+    PropertiesUpdated(topic, update, NT_EVENT_NONE, true, false);
+  }
+}
+
+void StorageImpl::SetPersistent(LocalTopic* topic, bool value) {
+  PropertiesUpdated(topic, topic->SetPersistent(value), NT_EVENT_NONE, true,
+                    false);
+}
+
+void StorageImpl::SetRetained(LocalTopic* topic, bool value) {
+  PropertiesUpdated(topic, topic->SetRetained(value), NT_EVENT_NONE, true,
+                    false);
+}
+
+void StorageImpl::SetCached(LocalTopic* topic, bool value) {
+  PropertiesUpdated(topic, topic->SetCached(value), NT_EVENT_NONE, true, false);
+}
+
+void StorageImpl::SetProperty(LocalTopic* topic, std::string_view name,
+                              const wpi::json& value) {
+  PropertiesUpdated(topic, topic->SetProperty(name, value), NT_EVENT_NONE,
+                    true);
+}
+
+bool StorageImpl::SetProperties(LocalTopic* topic, const wpi::json& update,
+                                bool sendNetwork) {
+  DEBUG4("SetProperties({},{})", topic->name, sendNetwork);
```

Comment:
Note that this debug will be emitted even if `update` is not an object. Not necessarily bad, but it is technically a change in behavior.

Reply:
It's a debug print, so that's fine.